### PR TITLE
Add initial basic support for vectorized forward mode AD

### PR DESCRIPTION
**Pull request by vaithak**
This PR adds initial support for vectorized forward mode for differentiating functions w.r.t all parameters in forward mode.

Minimal working example: 
```cpp
double f(double x, double y) {
  return x*y;
}

int main() {
  auto f_dx = clad::vector_forward_differentiate(f);
  double dx = 0, dy = 0;
  f_dx.execute(/*x=*/3, /*y=*/4, &dx, &dy);
  std::cout << "dx: " << dx << ' ' << "dy: " << dy << std::endl;
  
  // Dump the generated derivative code to standard output.
  f_dx.dump();
}
``` 

Corresponding output:
```bash
dx: 4 dy: 3
The code is: 
void f_d_all_args(double x, double y, double *_d_x, double *_d_y) {
    clad::array<double> _d_vector_x = {1., 0.};
    clad::array<double> _d_vector_y = {0., 1.};
    {
        clad::array<double> _d_vector_return = _d_vector_x * y + x * _d_vector_y;
        *_d_x = _d_vector_return[0];
        *_d_y = _d_vector_return[1];
        return;
    }
}
```

### CODE REVIEW

**Comment by vaithak** [comment]
Hi, I need some help with the falling test cases; the test is failing for Clang 15 because it is generating extra pragmas related to floating point arithmetic with the code:
```cpp
void f1_d_all_args(double x, double y, clad::array_ref<double> _d_x, clad::array_ref<double> _d_y) { 
   clad::array<double> _d_vector_x = {1, 0}; 
   clad::array<double> _d_vector_y = {0, 1}; 
   { 
        #pragma clang fp exceptions() 
        #pragma STDC FENV_ROUND FE_DYNAMIC 
        clad::array<double> _d_vector_return = _d_vector_x * y + x * _d_vector_y; 
        * _d_x = _d_vector_return[0]; 
        * _d_y = _d_vector_return[1]; 
    } 
} 
```
Do you have any ideas or suggestions as to what I am missing here and how this can be fixed?

**Comment by vgvassilev** [comment]
I remember we needed some adjustments in the recent clang versions for floating point arithmetic in our compatibility layer. Could you build llvm in debug and set a breakpoint at the time the pragma is produced. 

Cc: @alexander-penev

**Comment by vaithak** [comment]
Thanks for the help, @vgvassilev, using [the custom defined function](https://github.com/vgvassilev/clad/blob/master/include/clad/Differentiator/Compatibility.h#L64) fixed the tests 👍🏼

File: include/clad/Differentiator/Array.h [hunk 1]
```diff
@@ -192,6 +201,114 @@ template <typename T> class array {
       m_arr[i] /= arr[i];
     return *this;
   }
+  /// Multiplies the number to every element in the array and returns a new
+  /// array
+  CUDA_HOST_DEVICE array<T> operator*(T n) {
```
Comment (parth-07) [inline include/clad/Differentiator/Array.h:206]: Can it be a `const` member function?

File: include/clad/Differentiator/Array.h [hunk 2]
```diff
@@ -192,6 +201,114 @@ template <typename T> class array {
...
+    array<T> arr(m_size);
+    for (std::size_t i = 0; i < m_size; i++)
+      arr[i] = m_arr[i] * n;
+    return arr;
+  }
+  /// Multiplies the number to every element in the array and returns a new
+  /// array, when the number is on the left side
+  CUDA_HOST_DEVICE friend array<T> operator*(T n, array<T>& arr) {
```
Comment (parth-07) [inline include/clad/Differentiator/Array.h:214]: Can we pass `arr` as a const reference instead? If yes, then please make changes in all such operator overloads.

File: include/clad/Differentiator/Array.h [hunk 3]
```diff
@@ -192,6 +201,114 @@ template <typename T> class array {
...
+  CUDA_HOST_DEVICE friend array<T> operator+(array<T>& arr1, array<T>& arr2) {
+    assert(arr1.size() == arr2.size());
+    array<T> arr(arr1.size());
+    for (std::size_t i = 0; i < arr1.size(); i++)
+      arr[i] = arr1[i] + arr2[i];
+    return arr;
+  }
+  /// Friend function to define element wise adding of two arrays, one is
+  /// lvalue and other is rvalue.
+  CUDA_HOST_DEVICE friend array<T> operator+(array<T>& arr1, array<T>&& arr2) {
```
Comment (parth-07) [inline include/clad/Differentiator/Array.h:251]: Why do you have overloads that differ in the parameters `array<T> &arr2` and `array<T> &&arr2`?

File: include/clad/Differentiator/Array.h [hunk 4]
```diff
@@ -192,6 +201,114 @@ template <typename T> class array {
...
+  CUDA_HOST_DEVICE friend array<T> operator+(array<T>&& arr1, array<T>& arr2) {
+    assert(arr1.size() == arr2.size());
+    array<T> arr(arr1.size());
+    for (std::size_t i = 0; i < arr1.size(); i++)
+      arr[i] = arr1[i] + arr2[i];
+    return arr;
+  }
+  /// Friend function to define element wise adding of two arrays, both are
+  /// rvalues.
+  CUDA_HOST_DEVICE friend array<T> operator+(array<T>&& arr1, array<T>&& arr2) {
```
Comment (parth-07) [inline include/clad/Differentiator/Array.h:269]: Can you please explain why overloads for rvalue references are beneficial here? You are not moving any value
within the function body.

File: include/clad/Differentiator/Differentiator.h [hunk 5]
```diff
@@ -317,6 +320,22 @@ namespace clad {
...
+  /// Specialization for differentiating using vector mode.
+  /// The specialization is needed because the derivative function generated
+  /// by clad for vector mode is different from the one generated for forward
+  /// mode.
+  template <signed N = 1, typename ArgSpec = const char*, typename F,
+            typename DerivedFnType = ExtractDerivedFnTraitsVecForwMode_t<F>,
+            typename = typename std::enable_if<N==clad::opts::vector_fwd_mode &&
+                !std::is_class<remove_reference_and_pointer_t<F>>::value>::type>
+  CladFunction<DerivedFnType, ExtractFunctorTraits_t<F>, true> __attribute__((
+      annotate("D")))
```
Comment (parth-07) [inline include/clad/Differentiator/Differentiator.h:332]: `annotate("D")` is used to recognize the differentiation mode. We should have different differentiation mode for ordinary forward and vector forward mode.

**Review by parth-07 — CHANGES_REQUESTED** [review]
Initial review.

**Comment by parth-07** [comment]
@vaithak How can a user request for 4th order differentiation using vector forward mode? It seems that
`clad::opts::vector_fwd_mode` and differentiation order shares the same template parameter.

File: include/clad/Differentiator/Array.h (hunk 1 above)
Comment (vaithak) [reply include/clad/Differentiator/Array.h:206]: fixed 👍🏼

File: include/clad/Differentiator/Array.h (hunk 2 above)
Comment (vaithak) [reply include/clad/Differentiator/Array.h:214]: changed in all such operator overloads 👍🏼

File: include/clad/Differentiator/Array.h (hunk 4 above)
Comment (vaithak) [reply include/clad/Differentiator/Array.h:281]: Yeah, my bad. I have fixed this by making changes in the rvalue reference itself.

File: include/clad/Differentiator/Array.h (hunk 3 above)
Comment (vaithak) [reply include/clad/Differentiator/Array.h]: This is mainly used for cascaded operations. For ex. `clad::array<double> d = a+b+c`, where `a, b, c` are also clad arrays.

**Comment by vaithak** [comment]
Hi @parth-07, vector forward mode will calculate derivative w.r.t all parameters. Ideally, this should mean calculating the Hessian matrix for n=2 and similarly for other higher-order derivatives. So, I think N>1 won't make much sense for vector fwd mode.

File: include/clad/Differentiator/Differentiator.h (hunk 5 above)
Comment (vaithak) [reply include/clad/Differentiator/Differentiator.h:332]: commented below about the reasoning for this.

File: lib/Differentiator/ForwardModeVisitor.cpp [hunk 6]
```diff
@@ -203,6 +204,152 @@ namespace clad {
...
+
+    // Generate the function type for the derivative.
+    llvm::SmallVector<clang::QualType, 8> paramTypes;
+    paramTypes.reserve(m_Function->getNumParams() * 2);
+    for (auto PVD : m_Function->parameters()) {
+      paramTypes.push_back(PVD->getType());
+    }
+    for (auto PVD : m_Function->parameters()) {
+      QualType ValueType = utils::GetValueType(PVD->getType());
+      ValueType.removeLocalConst();
```
Comment (parth-07) [inline lib/Differentiator/ForwardModeVisitor.cpp:270]: What if the original type is a reference type?

File: lib/Differentiator/ForwardModeVisitor.cpp [hunk 7]
```diff
@@ -957,10 +1146,21 @@ namespace clad {
     if (auto VD = dyn_cast<VarDecl>(clonedDRE->getDecl())) {
       // If DRE references a variable, try to find if we know something about
       // how it is related to the independent variable.
-      auto it = m_Variables.find(VD);
-      if (it != std::end(m_Variables)) {
+      clang::Expr* dExpr = nullptr;
+      if (m_VectorMode) {
+        // If we are in vector mode, use the vector of derivatives.
+        auto it = m_VectorVariables.find(VD);
```
Comment (parth-07) [inline lib/Differentiator/ForwardModeVisitor.cpp:1152]: Is there any use of `m_Variables` in vector forward mode?

**Review by parth-07 — CHANGES_REQUESTED** [review]

**Comment by parth-07** [comment]
> Hi @parth-07, vector forward mode will calculate derivative w.r.t all parameters. This should mean calculating the Hessian matrix for n=2 and similarly for other higher-order derivatives. So, I think N>1 won't make much sense for vector fwd mode. 

Hi @vaithak
Yes, vector forward mode will compute derivatives w.r.t all the parameters. For a second-order differentiation, wouldn't this mean computing the diagonal of the Hessian matrix instead of the matrix itself? We already have a requirement for such a case: issue #509. 

We shouldn't use the same function template parameter for entirely different use cases, even if we do not wish to support higher-order differentials for vector forward mode. We can use a different function altogether for vector forward mode, `clad::VectorForwardDifferentiation`. @vgvassilev @vaithak  What do you think about this?

File: include/clad/Differentiator/Array.h [hunk 8]
```diff
@@ -192,6 +201,117 @@ template <typename T> class array {
...
+    for (std::size_t i = 0; i < m_size; i++)
+      arr[i] = m_arr[i] * n;
+    return arr;
+  }
+  /// Multiplies the number to every element in the array and returns a new
+  /// array, when the number is on the left side
+  CUDA_HOST_DEVICE friend array<T> operator*(T n, const array<T>& arr) {
+    return arr * n;
+  }
+  CUDA_HOST_DEVICE friend array<T> operator*(T n, array<T>&& arr) {
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply include/clad/Differentiator/Array.h:217]: Yes, please run `git-clang-format` to this patch.

File: include/clad/Differentiator/Array.h (hunk 3 above)
Comment (vaithak) [reply include/clad/Differentiator/Array.h]: removed all rvalue reference functions for now to keep this simple, will add them later if required.

File: include/clad/Differentiator/Array.h (hunk 4 above)
Comment (vaithak) [reply include/clad/Differentiator/Array.h]: removed rvalue reference function 👍🏼

File: include/clad/Differentiator/Array.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (vaithak) [reply include/clad/Differentiator/Array.h]: ran `git-clang-format` and added spaces between functions 👍🏼

File: include/clad/Differentiator/Array.h [hunk 9]
```diff
@@ -192,6 +201,117 @@ template <typename T> class array {
...
+  CUDA_HOST_DEVICE array<T> operator+(T n) const {
+    array<T> arr(m_size);
+    for (std::size_t i = 0; i < m_size; i++)
+      arr[i] = m_arr[i] + n;
+    return arr;
+  }
+  CUDA_HOST_DEVICE friend array<T> operator+(T n, const array<T>& arr) {
+    return arr + n;
+  }
+  CUDA_HOST_DEVICE friend array<T> operator+(T n, array<T>&& arr) {
```
[the remark this answers is not in the document]
Comment (vaithak) [reply include/clad/Differentiator/Array.h]: removed 👍🏼

File: include/clad/Differentiator/ForwardModeVisitor.h [hunk 10]
```diff
@@ -96,6 +102,20 @@ namespace clad {
...
+    //  vectorized forward mode.
+    ///
+    /// Information about the original function, derived function, derived
+    /// function parameter types and the differentiation mode are implicitly
+    /// taken from the data member variables. In particular, `m_Function`,
+    /// `m_Mode` and `m_Derivative` should be correctly set before using this
+    /// function.
+    llvm::SmallVector<clang::ParmVarDecl*, 8> BuildVectorModeParams(DiffParams& diffParams);
+
+    /// Get an expression used to initialize the one-hot vector for the
```
[the remark this answers is not in the document]
Comment (vaithak) [reply include/clad/Differentiator/ForwardModeVisitor.h:117]: explained in comments 👍🏼

File: lib/Differentiator/ForwardModeVisitor.cpp [hunk 11]
```diff
@@ -203,6 +204,152 @@ namespace clad {
...
+                                              PVD->getStorageClass());
+      paramDerivatives.push_back(dPVD);
+      ++dParamTypesIdx;
+
+      if (dPVD->getIdentifier())
+            m_Sema.PushOnScopeChains(dPVD, getCurrentScope(),
+                                     /*AddToContext=*/false);
+
+      m_Variables[newPVD] = BuildOp(UO_Deref, BuildDeclRef(dPVD),
+                                       m_Function->getLocation());
```
[the remark this answers is not in the document]
Comment (vaithak) [reply lib/Differentiator/ForwardModeVisitor.cpp]: passed `noLoc` 👍🏼

File: lib/Differentiator/ForwardModeVisitor.cpp (hunk 6 above)
Comment (vaithak) [reply lib/Differentiator/ForwardModeVisitor.cpp:273]: wouldn't `utils::GetValueType` take care of that?

File: lib/Differentiator/ForwardModeVisitor.cpp (hunk 7 above)
Comment (vaithak) [reply lib/Differentiator/ForwardModeVisitor.cpp:1152]: yup, `m_Variables` is for storing the mapping from original function parameters to new parameters for derivatives.

**Comment by vaithak** [comment]
> > Hi @parth-07, vector forward mode will calculate derivative w.r.t all parameters. This should mean calculating the Hessian matrix for n=2 and similarly for other higher-order derivatives. So, I think N>1 won't make much sense for vector fwd mode.
> 
> Hi @vaithak Yes, vector forward mode will compute derivatives w.r.t all the parameters. For a second-order differentiation, wouldn't this mean computing the diagonal of the Hessian matrix instead of the matrix itself? We already have a requirement for such a case: issue #509.
> 
> We shouldn't use the same function template parameter for entirely different use cases, even if we do not wish to support higher-order differentials for vector forward mode. We can use a different function altogether for vector forward mode, `clad::VectorForwardDifferentiation`. @vgvassilev @vaithak What do you think about this?

I think this makes sense, will think more about this and let you know if I find some edge cases otherwise will add the new method for vector mode as you suggested.

**Comment by vaithak** [comment]
@parth-07 I have created a new function for `clad::vector_forward_differentiate` instead of having an `opts` 👍🏼

File: lib/Differentiator/ForwardModeVisitor.cpp [hunk 12]
```diff
@@ -221,6 +368,13 @@ namespace clad {
     if (DVI.empty())
       return {};
 
+    if (request.vector_fwd_mode) {
+      assert (DVI.size() == FD->getNumParams() &&
+              "Vector mode supports differentiating w.r.t. all parameters only.");
```
[the remark this answers is not in the document]
Comment (vaithak) [reply lib/Differentiator/ForwardModeVisitor.cpp]: created an issue: https://github.com/vgvassilev/clad/issues/566

File: lib/Differentiator/ForwardModeVisitor.cpp (hunk 6 above)
Comment (parth-07) [reply lib/Differentiator/ForwardModeVisitor.cpp:276]: Oh, yes, It would.

File: include/clad/Differentiator/Array.h [hunk 13]
```diff
@@ -192,6 +200,78 @@ template <typename T> class array {
...
+  }
+  CUDA_HOST_DEVICE friend array<T> operator-(T n, const array<T>& arr) {
+    array<T> arr2(arr.size());
+    for (std::size_t i = 0; i < arr.size(); i++)
+      arr2[i] = n - arr[i];
+    return arr2;
+  }
+
+  /// Friend function to define element wise adding of two arrays.
+  CUDA_HOST_DEVICE friend array<T> operator+(const array<T>& arr1,
```
Comment (parth-07) [inline include/clad/Differentiator/Array.h:258]: Why does it need to be a friend function? It does not seem to be accessing any private/protected members.
Also, what do you think about moving the operators outside the class?

File: include/clad/Differentiator/FunctionTraits.h [hunk 14]
```diff
@@ -719,6 +719,25 @@ namespace clad {
   using ExtractDerivedFnTraitsForwMode_t =
       typename ExtractDerivedFnTraitsForwMode<F>::type;
 
+  template <class T, class R> struct OutputVecParamType {
+    using type = typename std::add_pointer<R>::type;
```
Comment (parth-07) [inline include/clad/Differentiator/FunctionTraits.h:723]: Will it be correct if the return type is a `const` type?

**Review by parth-07 — CHANGES_REQUESTED** [review]

File: include/clad/Differentiator/Array.h [hunk 15]
```diff
@@ -40,6 +41,13 @@ template <typename T> class array {
 
   CUDA_HOST_DEVICE array(const array<T>& arr) : array(arr.m_arr, arr.m_size) {}
 
+  CUDA_HOST_DEVICE array(std::initializer_list<T> arr)
+      : m_arr(new T[arr.size()]{static_cast<T>(T())}), m_size(arr.size()) {
```
[the remark this answers is not in the document]
Comment (vaithak) [reply include/clad/Differentiator/Array.h:45]: I followed the convention as it is done for other constructors of this class.

File: include/clad/Differentiator/Array.h (hunk 13 above)
Comment (vaithak) [reply include/clad/Differentiator/Array.h]: shifted most of the overloaded operators outside the class.

File: include/clad/Differentiator/FunctionTraits.h (hunk 14 above)
Comment (vaithak) [reply include/clad/Differentiator/FunctionTraits.h:723]: yup, it works even if return type is `const double` for example

File: test/ForwardMode/VectorMode.C [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+//CHECK-NOT: {{.*error|warning|note:.*}}
+
+#include "clad/Differentiator/Differentiator.h"
+#include <cmath>
+
+double f1(double x, double y) {
+  return x*y;
+}
+
+void f1_d_all_args(double x, double y, double *_d_x, double *_d_y);
```
Comment (vgvassilev) [inline test/ForwardMode/VectorMode.C:13]: Could we extend the test cases? We should probably add a demo (in the demos folder). We should also document the new feature and interface in the documentation under `docs`.

Comment (vgvassilev) [inline test/ForwardMode/VectorMode.C:13]: Could we extend the test cases? We should probably add a demo (in the demos folder). We should also document the new feature and interface in the documentation under `docs`.

Comment (vaithak) [reply test/ForwardMode/VectorMode.C:13]: Sure, I will add it in follow-up PRs when the implementation is a little stable for some decently complex programs.

Comment (vgvassilev) [reply test/ForwardMode/VectorMode.C:13]: According to codecov this PR needs more tests to reach 100% coverage. I don’t think we need complex examples but examples good enough to test all new code that we have.

Comment (vaithak) [reply test/ForwardMode/VectorMode.C:12]: added more test cases 👍🏼

File: include/clad/Differentiator/Array.h [hunk 17]
```diff
@@ -192,9 +200,100 @@ template <typename T> class array {
...
-};
+}; // class array
+
+/// Overloaded operators for clad::array which return a new array.
+
+/// Multiplies the number to every element in the array and returns a new
+/// array.
+template <typename T>
+CUDA_HOST_DEVICE array<T> operator*(const array<T>& arr, T n) {
+  array<T> arr2(arr);
```
Comment (vgvassilev) [inline include/clad/Differentiator/Array.h:232]: Do we need an extra allocation here? Can we use `array_ref`?

File: include/clad/Differentiator/Array.h [hunk 18]
```diff
@@ -192,9 +200,100 @@ template <typename T> class array {
...
+CUDA_HOST_DEVICE array<T> operator/(const array<T>& arr, T n) {
+  array<T> arr2(arr);
+  arr2 /= n;
+  return arr2;
+}
+
+/// Adds the number to every element in the array and returns a new array
+template <typename T>
+CUDA_HOST_DEVICE array<T> operator+(const array<T>& arr, T n) {
+  array<T> arr2(arr);
```
Comment (vgvassilev) [inline include/clad/Differentiator/Array.h:256]: Likewise.

File: include/clad/Differentiator/Array.h (hunk 17 above)
Comment (vaithak) [reply include/clad/Differentiator/Array.h:232]: The objective of this function is to return a new array (with multiplied entries),  if we use array_ref, won't `arr`'s underlying array also get modified? 
Also, if a new array is needed to be returned, won't it anyway require allocating a new array?

**Review by parth-07 — APPROVED** [review]
Overall looks good. Can you please squash all the commits into one?

**Comment by vaithak** [comment]
created a single commit 👍🏼

File: include/clad/Differentiator/Array.h (hunk 17 above)
Comment (vaithak) [reply include/clad/Differentiator/Array.h:232]: This overloaded operator is outside the class definition, so we can't use `m_arr` here.

Comment (vaithak) [reply include/clad/Differentiator/Array.h:232]: @vgvassilev, can we mark this as resolved?

Comment (vgvassilev) [reply include/clad/Differentiator/Array.h:232]: Yep, thanks for clarifying!

**Pull request merged**